### DISCUSSION TITLE
SSD1309 Oled support

### DISCUSSION THREAD

**Discussion by martin072** [arendst/Tasmota#19605] [Q&A]
Quick question does tasmota support oled displays with a SSD1309 chipset? (I2c)

**Comment by sfromis** [comment]
I saw someone on the net mention that SSD1309 should be usable based on SSD1306 support, which Tasmota has. No experience with this.

More generally, support for new display types, or variants, are not done as Tasmota core code, but in a separate display descriptor configuration file.

**Reply by martin072** [reply]
In the my_user_config, it specify's ```USE_DISPLAY_SSD1306``` and I also notice an adafruit library is included in tasmota, so that seems to be supported in code? there some ssd1309 library's around as well.

it still a 128x64 display, but slightly bigger (1.54") when I hook it up it gives a garbled screen (not using lvgl, but the standard Tasmota displayMode / displayText commands). so not sure if it is that compatible with an SSD1306

**Reply by sfromis** [reply]
Well, that "someone" said that it is compatible, having compared the commands. Thus, based on the SSD1306 support, it should work, based on those statements.

Using universal display is not specifically for LVGL/HASPmota, but also DisplayText commands, the idea is to **not** be having Tasmota core code being changed for each display type.

**Reply by martin072** [reply]
Well, cleary it doesn't work 😅 as a simple swap, so either the chinese lie about the compatibility or there is more to it. Nevertheless thank you for the time

**Reply by sfromis** [reply]
Did you make sure to have the I2C address detected, and matching what you have in DisplayAddess?

**Reply by martin072** [reply]
Yes, when tasmota starts, it finds a 'SSD1306' display at 0x3c it says:

```
00:00:00.268 CFG: Energy use defaults as file system not ready or file not found
00:00:00.268 I2C: SSD1306 found at 0x3c
00:00:00.295 DSP: SSD1306
00:00:00.303 Project tasmota - TASMOTA Version 13.0.0(tasmota32)-2_0_10(2023-09-19T16:24:42)
```

**Reply by sfromis** [reply]
And did you also verify that it matches DisplayAddress? That would be 60 in decimal.

(You also need to remember to power on the display, not just config it).

**Reply by martin072** [reply]
Yes it is ``` 13:42:23.505 RSL: RESULT = {"DisplayAddress1":60}```
As mentioned, a older 0.96" works fine, just the new 'SSD1309', so not so compatible as advertised, as I assume nothing has to be changed if it is compatible?

See pics
![IMG_7954](https://github.com/arendst/Tasmota/assets/24524506/ec96c8a3-01f8-4468-98bc-16511a26ce4d)
![IMG_7955](https://github.com/arendst/Tasmota/assets/24524506/47bd74dd-6090-4460-80af-3be7426e6b0f)
![IMG_7956](https://github.com/arendst/Tasmota/assets/24524506/097f4b92-aa46-4e2f-b73f-8c31d74111bd)

**Comment by martin072** [comment]
Yes, ```20:17:01.389 RSL: RESULT = {"DisplayAddress1":60}```

See pictures what happens when i just 'swap' the displays (assuming the are compatible as advertised)

![IMG_7954](https://github.com/arendst/Tasmota/assets/24524506/3d4d19b2-6628-439a-aaac-ee54034aa3d3)
![IMG_7955](https://github.com/arendst/Tasmota/assets/24524506/12c47e67-ed64-4d09-8c48-9d127bdde067)
![IMG_7956](https://github.com/arendst/Tasmota/assets/24524506/1fdfe9ad-dc00-44c7-adb8-78a214ce5a7a)

**Reply by Jason2866** [reply]
The PCB side photos are the from the working and non working one?

**Reply by martin072** [reply]
That's from the non working one

**Reply by Jason2866** [reply]
The working one would be interesting for compare

**Comment by martin072** [comment]
This is the working one 

![IMG_7990](https://github.com/arendst/Tasmota/assets/24524506/3f6aca3d-a407-4e36-a0c7-3ba472aca002)

**Comment by Jason2866** [comment]
Looks like there is a selection done with the resistors to use i2c or 3 or 4 wire SPI.
Search if you find "somewhere" the setting which is used for the working one and modify the not working one to this setup. 
The datasheet where it is described which IO pins needs to be set to enable i2c mode
https://www.buydisplay.com/download/ic/SSD1309.pdf

**Reply by martin072** [reply]
Will have a look, it came in two choices, i2c -or spi, as the pin layout clearly is for i2c, one would assume correct selection is already made, otherwise it would be sold in a non working state?

**Reply by Jason2866** [reply]
Good one. Do you think the Chinese sellers take care off such details?

**Reply by martin072** [reply]
Had some time to check, and it works fine on both display's with another library configured as I2C (https://github.com/olikraus/u8g2)
Used the 8x8->GraphicDemo example and configured as ```U8X8_SSD1306_128X64_NONAME_HW_I2C u8x8(/* reset=*/ U8X8_PIN_NONE);```

(Part of the font missing is caused by photographing it with the phone, display is fine)

![IMG_8050](https://github.com/arendst/Tasmota/assets/24524506/ea78f960-98ca-4522-8547-522d8efe414b)
![IMG_8049](https://github.com/arendst/Tasmota/assets/24524506/a66f04f1-e594-4d71-8f40-596219c481e2)

**Comment by hk2k2** [comment]
Hello! 
Have you tried it using Universal Display Driver?
Just tried a new bought SSD1309 with my existing udisplay driver config for SSD1306, works fine.
![ssd1309](https://github.com/arendst/Tasmota/assets/70974088/a16dcb7e-159a-41a1-997c-571e1fef998d)
![ssd1309_back](https://github.com/arendst/Tasmota/assets/70974088/cdea466e-3b2c-4b11-a222-98368f7e14b3)


also tried a complete new flashed ESP with SSD1306 compiled (no udisplay), works fine too! 
tried with 13.3.0.4 ..

it is a 2.42" version, I2c Adress is 0x3C -- display was SPI version and i soldered it to I2C

BR hk2k2

**Reply by Gittegatt** [reply]
@hk2k2 
Thank you for providing these pictures.
Possible to take another photo of the resistors underneath the flat ribbon cable?

 Cheers.

**Reply by sfromis** [reply]
This thread has a decent image of what's under the ribbon:
https://forum.arduino.cc/t/2-42-oled-conversion/1099068

**Reply by hk2k2** [reply]
![ssd1309_back](https://github.com/arendst/Tasmota/assets/70974088/5cc1b771-282e-47cf-9c56-169a27585355)
i just soldered the R11 0R bridge, all others are default underneath the flat cable..

**Reply by Gittegatt** [reply]
@hk2k2 
Thank you. Would just like to verify if I've done everything properly to be ready for I²C. I know that I've ran the display with I²C some month ago (no tasmota), but would like to confirm my assumption. So, I tried the display run with tasmota 3 days ago but had no luck. Then, 2 days ago everything worked fine out of the box. I don't know what I've done in other ways.
Thank you, anyway.

Regards,
Daniel.

**Comment by DennisXK** [comment]
Hello, has anyone now gotten the 2.42" OLED to work with tasmota? The statements above are not clear.

https://de.aliexpress.com/item/1005006349198160.html?spm=a2g0o.productlist.main.57.6c4e568363q7io&algo_pvid=dbf8feb6-4335-4fd9-bb88-b30f8856aaeb&algo_exp_id=dbf8feb6-4335-4fd9-bb88-b30f8856aaeb-28&pdp_npi=4%40dis%21EUR%219.99%217.49%21%21%2174.72%2156.02%21%402103835c17315644401834498ee7c4%2112000036848827498%21sea%21DE%211878639479%21X&curPageLogUid=o9rOqNRzdwaF&utparam-url=scene%3Asearch%7Cquery_from%3A

**Reply by Gittegatt** [reply]
Hi,
mine is running with Tasmota 13.4.0
Some month ago I was about to update to a higher release but I did not get any output on the display. So I decided to stay with 13.4.0

**Comment by DennisXK** [comment]
ok thx, and which version did you use exactly? Can I use it from the link? I2C or SPI version?

**Comment by Gittegatt** [comment]
Yes, you can use the one provided in your link.
In general the board has the ability to use both interfaces - i2c and SPI. You can find the preference on the backside of the PCB to choose from.
But if you would like to go safe, choose the i2c or SPI with the description in the offering.

**Comment by james-lan** [comment]
Tasmota works with SSD1309, and it took a bit of work to get done: 

Specific board is this: http://www.lcdwiki.com/2.42inch_SPI_OLED_Module_SKU%3AMSP242X



Using i2c, Which meant desoldering R5, and bridging R9 (resistor from R5) & R4 (piece of wire) 
Note, R8 should be bridged, otherwise CS needs low. (WAS NOT DONE, THUS TX pin below)

Directed to a LCD->D1 mini (Tasmota Setting):
GND -> G (Not Applicable)
VCC -> D4 (Output Hi)
SCL -> D3 (I2C SCL)
SDA -> D2 (I2C SDA)
RES -> D1 (None)
DC -> RX (None)
CS -> TX (Output Lo OR None (If D8 was soldered))

Set another pin, I chose D6 (as it's empty right now) As Option A  3

Rule3 :H,SSD1306,128,64,1,I2C,3d,*,*,* :S,0,2,1,0,30,20  :O AE AF :I AE AF  :A,00,10,40,00,00 :i,A6,A7 #

Reboot
DisplayText Tasmota[l2c1]Supports[l3c1]SSD1309[l4c1][T]

NOTE: If you include :o AE, it will cause it to reset, and sometimes without the :O having AE [Reset] AF [ON] it will not update. So for the most part it's the same as SSD1306, but there are probably still some differences, but hey this works with 14.5.0 (-display) on a D1 mini 8266
